Add support for non-YCbCr input to ToI420

lherman-cs opened the pull request:
Resolves https://github.com/pion/mediadevices/issues/46

at-wat commented on pkg/io/video/convert.go:
```diff
@@ -3,8 +3,38 @@ package video
 
+// imageToYCbCr converts image to *image.YCbCr
+// Note: conversion can be lossy
+func imageToYCbCr(img image.Image) *image.YCbCr {
```
It would be better to receive the destination pointer as an argument to reuse allocated memory, and realloc Y, Cb, Cr slices if the size of the image is changed.

lherman-cs replied:
Makes senses, I'll update it

at-wat commented on pkg/io/video/convert.go:
```diff
@@ -3,8 +3,38 @@ package video
+			yuvImg.Y[i] = yy
+			yuvImg.Cb[i] = cb
+			yuvImg.Cr[i] = cr
+			i++
```
Input image can be padded.
It's safer to do like
https://github.com/pion/mediadevices/blob/cad9f410477aa2611c18940d54336a56121272d8/pkg/driver/screen/x11_linux.go#L77-L90
(This block can be removed in this PR!)

lherman-cs replied:
hmm, not sure what you mean by padded. Do you mean about `YStride` and `CStride`?

at-wat replied:
Oh, sorry. This image is allocated by us, so we don't have to take padding of the image into account.
(I meant for example `Width=15, YStride=16 (4-byte aligned, 1-byte padded), Rect.Max.X = 15`)

lherman-cs replied:
Yep, we have control over this. And, to make it simple, we'll just always convert it to I444.

lherman-cs commented:
Updated PR with some fixes to the transform example. Also, changed the video transform function to use `imaging.Rotate180`, this will show the users how flexible the API too.

lherman-cs commented:
@at-wat Not sure what I did wrong. But, the color image got messed up...
![Screenshot from 2020-02-15 00-16-29](https://user-images.githubusercontent.com/15654932/74584547-85f82480-4f88-11ea-942d-afbcdd6da041.png)

at-wat commented on pkg/io/video/convert.go:
```diff
@@ -3,18 +3,66 @@ package video
+
+	bounds := src.Bounds()
+	dy := bounds.Max.Y - bounds.Min.Y
+	dx := bounds.Max.X - bounds.Min.X
```
Just FYI
```suggestion
	dx := bounds.Dx()
```

lherman-cs replied:
I didn't know that method existed... Thx

at-wat commented:
The color looks also broken in master branch.
Maybe I had only seen the monochrome terminal during testing https://github.com/pion/mediadevices/pull/42....

at-wat commented:
It's a bug in screen capture adapter and color conversion.
I made a PR https://github.com/pion/mediadevices/pull/56 to fix it.

lherman-cs commented:
Rebased and updated the PR to address the comments.

at-wat commented on examples/screenshare/main.go:
```diff
@@ -52,7 +52,7 @@ func main() {
 		Video: func(c *mediadevices.MediaTrackConstraints) {
 			c.CodecName = videoCodecName
-			c.FrameFormat = frame.FormatI444
+			c.FrameFormat = frame.FormatI420
```
This prop is sent to the screen capture driver and just ignored.

lherman-cs replied:
Ah. I'll remove it to make it clear. Also, I'll add a new const, `FormatRGBA` to correctly label the output from `Properties`

at-wat commented on pkg/driver/screen/x11_linux.go:
```diff
@@ -61,40 +60,16 @@ func (s *screen) Close() error {
+		return s.reader.Read(), nil
 	})
-	return r, nil
+	return video.ToI420(r), nil
```
I think it's better to insert ToI420 just before the encoder. RGBA image is good at manipulation as the image processing libraries have optimized implementation for RGBA image.

lherman-cs replied:
Ah right, that makes sense.

lherman-cs replied:
So, I've updated it, but I got a panic, https://gist.githubusercontent.com/lherman-cs/02234dd460897baa520cffc111460fe2/raw/ccc32139b4a84b8dc4c80b2bc2cc8354d53051ac/gistfile1.txt.

lherman-cs replied:
It looks like we have a bug in `vpx`. When I ran the example with `GODEBUG=cgocheck=2`, it crashed immediately and showed:

```
write of Go pointer 0xc000562000 to non-Go memory 0x12229e0

...
goroutine 27 [running]:
runtime.systemstack_switch()
        /usr/lib/go-1.13/src/runtime/asm_amd64.s:330 fp=0xc000073dc0 sp=0xc000073db8 pc=0x47b3d0
runtime.cgoCheckWriteBarrier(0x12229e0, 0xc000562000)
        /usr/lib/go-1.13/src/runtime/cgocheck.go:53 +0xc9 fp=0xc000073df8 sp=0xc000073dc0 pc=0x427ec9
runtime.wbBufFlush(0x12229e0, 0xc000562000)
        /usr/lib/go-1.13/src/runtime/mwbbuf.go:196 +0x83 fp=0xc000073e28 sp=0xc000073df8 pc=0x44caa3
runtime.gcWriteBarrier(0xc0002e97a0, 0xa11f60, 0xc00013d300, 0x0, 0x0, 0x0, 0x400, 0xc000346000, 0x99c301, 0x56293c, ...)
        /usr/lib/go-1.13/src/runtime/asm_amd64.s:1444 +0xae fp=0xc000073eb0 sp=0xc000073e28 pc=0x47d55e
github.com/pion/mediadevices/pkg/codec/vpx.(*encoder).Read(0xc000139720, 0xc000346000, 0x400, 0x400, 0x0, 0x0, 0x0)
        /home/lukas/Documents/pion/mediadevices/pkg/codec/vpx/vpx.go:135 +0x59a fp=0xc000073f78 sp=0xc000073eb0 pc=0x7c61ca
github.com/pion/mediadevices.(*videoTrack).start(0xc00033a180)
        /home/lukas/Documents/pion/mediadevices/track.go:139 +0x83 fp=0xc000073fd8 sp=0xc000073f78 pc=0x75cca3
runtime.goexit()
        /usr/lib/go-1.13/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000073fe0 sp=0xc000073fd8 pc=0x47d4a1
created by github.com/pion/mediadevices.newVideoTrack
        /home/lukas/Documents/pion/mediadevices/track.go:130 +0x331
...
```

at-wat replied:
It looks a pointer handling problem. https://github.com/pion/mediadevices/pull/60 should fix it.

at-wat commented:
I made some optimization based on this branch. https://github.com/pion/mediadevices/pull/62

lherman-cs commented:
@at-wat Just updated the PR to remove RGBA conversion. PTAL
